README.md

[![pub package](https://img.shields.io/pub/v/sticky_grouped_list.svg)](https://pub.dev/packages/sticky_grouped_list)
[![package publisher](https://img.shields.io/pub/publisher/sticky_grouped_list.svg)](https://pub.dev/packages/sticky_grouped_list)
![build](https://github.com/Dimibe/sticky_grouped_list/actions/workflows/main.yaml/badge.svg??branch=main)
 
A `ListView` in which list items can be grouped to sections. Based on [scrollable_positioned_list](https://pub.dev/packages/scrollable_positioned_list), which enables programatically scrolling of the list.

<img src="https://raw.githubusercontent.com/Dimibe/sticky_grouped_list/master/assets/new-screenshot-for-readme.png" width="300"> <img src="https://raw.githubusercontent.com/Dimibe/sticky_grouped_list/master/assets/chat.png" width="300">

### Features
* Easy creation of chat-like interfaces. 
* List items can be separated in groups.
* For the groups an individual header can be set.
* Sticky headers with floating option. 
* All fields from `ScrollablePositionedList` available.

## Getting Started

 Add the package to your pubspec.yaml:

 ```yaml
 sticky_grouped_list: ^3.1.0
 ```
 
 In your dart file, import the library:

 ```Dart
import 'package:sticky_grouped_list/sticky_grouped_list.dart';
 ``` 
 
 Create a `StickyGroupedListView` Widget:
 
 ```Dart
  final GroupedItemScrollController itemScrollController = GroupedItemScrollController();

  StickyGroupedListView<dynamic, String>(
    elements: _elements,
    sort:true,
    groupBy: (dynamic element) => element['group'],
    groupSeparatorBuilder: (dynamic element) => Text(element['group']),
    itemBuilder: (context, dynamic element) => Text(element['name']),
    itemComparator: (e1, e2) => e1['name'].compareTo(e2['name']), // optional
    elementIdentifier: (element) => element.name // optional - see below for usage
    itemScrollController: itemScrollController, // optional
    order: StickyGroupedListOrder.ASC, // optional
  );
```

If you are using the `GroupedItemScrollController` you can scroll or jump to an specific position in the list programatically:

1. By using the index, which will scroll to the element at position [index]:
```dart
  itemScrollController.scrollTo(index: 4, duration: Duration(seconds: 2));
  itemScrollController.jumpTo(index: 4);
```

2. By using a pre defined element identifier. The identifier is defined by a `Function` which takes one element and returns a unique identifier of any type.
The methods `scrollToElement` and `jumpToElement` can be used to jump to an element by providing the elements identifier instead of the index: 
```dart
  final GroupedItemScrollController itemScrollController = GroupedItemScrollController();

  StickyGroupedListView<dynamic, String>(
    elements: _elements,
    elementIdentifier: (element) => element.name
    itemScrollController: itemScrollController, 
    [...]
  );


  final scrolledItemIndex = itemScrollController.scrollToElement(identifier: 'item-1', duration: Duration(seconds: 2));
  final scrolledItemIndexOnJump = itemScrollController.jumpToElement(identifier: 'item-2');

```


### Parameters:
| Name                                 | Description                                                                                                                                                             | Required | Default value                |
|--------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----|------------------------------|
| `elements`                           | A list of the data you want to display in the list                                                                                                                      | required | -                            |
| `itemBuilder` / `indexedItemBuilder` | Function which returns an Widget which defines the item. `indexedItemBuilder` provides the current index as well. If both are defined `indexedItemBuilder` is preferred | yes, either of them | -                            |
| `groupBy`                            | Function which maps an element to its grouped value                                                                                                                     | required | -                            |
| `groupSeparatorBuilder`              | Function which gets a element and returns an Widget which defines the group header separator                                                                            | required | -                            |
| `separator`                          | A Widget which defines a separator between items inside a group                                                                                                         | no | no separator                 |
| `floatingHeader`                     | When set to `true` the sticky header will float over the list                                                                                                           | no | `false`                      |
| `stickyHeaderBackgroundColor`        | Defines the background color of the sticky header                                                                                                                       | no | `Color(0xffF7F7F7)`          |
| `itemScrollController`               | Instead of an `ItemScrollController` a `GroupedItemScrollController` needs to be provided.                                                                              | no | -                            |
| `elementIdentifier`                  | Used by `itemScrollController` and defines the unique identifier for each element.                                                                                      | no | -                            |
| `order`                              | Change to `StickyGroupedListOrder.DESC` to reverse the group sorting                                                                                                    | no | `StickyGroupedListOrder.ASC` |
| `groupComparator`                    | Can be used to define a custom sorting for the groups. Otherwise the natural sorting order is used                                                                      | no | -                            |
| `itemComparator`                     | Can be used to define a custom sorting for the elements inside each group. Otherwise the natural sorting order is used                                                  | no | -                            |
| `reverse`                            | Scrolls in opposite from reading direction (Starting at bottom and scrolling up). Same as in scrollable_positioned_list.                                                | no | false                        |
| `sort`                               | Enables sorting by given order.                                                                                                                                         | no | true                         |

*`GroupedItemScrollController.srollTo()` and `GroupedItemScrollController.jumpTo()` automatic set the `alignment` so that the item is fully visible aligned under the group header. Both methods take `automaticAlignment` as a additional optional paramenter which needs to be set to true if `alignment` is specified.*

**Also the fields from `ScrollablePositionedList.builder` can be used.**

## Highlight - Chat Dialog

Easy creation of chat-like dialogs.
Just set the option `reverse` to `true` and the option `order` to `StickyGroupedListOrder.DESC`. A full example can be found in the examples.
The list will be scrolled to the end in the initial state and therefore scrolling will be against redeaing direction. 

## Difference between grouped_list and sticky_grouped_list: 

TThe list views in the [GroupedList](https://pub.dev/packages/grouped_list) package are based on the default flutter listview and the silver list. This package is based on the [scrollable_positioned_list](https://pub.dev/packages/scrollable_positioned_list) which enables the possibility to programatically scroll to certain positions in the list. So if you need the ability to programatically scroll the list use the this package otherwise I would recommend to use the [GroupedList](https://pub.dev/packages/grouped_list) package.


## Used packages: 
| Package name | Copyright | License |
|----|----|----|
|[scrollable_positioned_list](https://pub.dev/packages/scrollable_positioned_list) | Copyright 2018 the Dart project authors, Inc. All rights reserved | [BSD 3-Clause "New" or "Revised" License](https://github.com/Dimibe/sticky_grouped_list/blob/master/LICENSE) |


CHANGELOG.md
## [3.1.1] - 04.09.2023

* Add param `sort`

## [3.1.0] - 07.07.2022

* Support for jump/scroll to element
* Fix initialScrollIndex
* Expose ItemPositionsListener

## [3.0.0] - 18.06.2022

* Dart 2.17.0
* Flutter lints and version upgrades

## [2.1.0] - 04.02.2022

* Support Shrink Wrap

## [2.0.0] - 04.02.2022

* Sound null-safety
* Updated dependencies

## [1.3.0] - 15.11.2020

* Update dependencies

## [1.2.0] - 19.09.2020

* New feature custom sorting. Added two new options `groupComparator` and `itemComparator`which can
  be used for comparison between two groups and items. If used the functions will be used for
  sorting the list.

## [1.1.0] - 06.09.2020

* New feature: The background color of the sticky header is now customizable (only if
  floatingHeaders option is not used).

## [1.0.0] - 05.09.2020

* New feature: Reverse list - easy chat dialog creation

## [0.3.0] - 25.07.2020

* Fix performance issue: Don't rebuild widget when sticky header changes.

## [0.2.1] - 04.07.2020

* Add code documentation
* improved README

## [0.2.0] - 30.06.2020

* scrollTo/jumpTo will set the element under the group header

## [0.1.2] - 26.06.2020

* Refactor example

## [0.1.1] - 24.06.2020

* Add example

## [0.1.0] - 23.06.2020

* Initial release

test/sticky_grouped_list_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:sticky_grouped_list/sticky_grouped_list.dart';

final List _elements = [
  {'name': 'John', 'group': 'Team A'},
  //{'name': 'Will', 'group': 'Team B'},
  // {'name': 'Beth', 'group': 'Team A'},
  {'name': 'Miranda', 'group': 'Team B'},
  // {'name': 'Mike', 'group': 'Team C'},
  {'name': 'Danny', 'group': 'Team C'},
];
void main() {
  Widget buildGroupSeperator(dynamic element) {
    return Text(element['group']);
  }

  testWidgets('find elemets and group separators', (WidgetTester tester) async {
    await tester.pumpWidget(
      MaterialApp(
        home: Scaffold(
          body: StickyGroupedListView(
            groupBy: (dynamic element) => element['group'],
            elements: _elements,
            order: StickyGroupedListOrder.DESC,
            groupSeparatorBuilder: buildGroupSeperator,
            itemBuilder: (context, dynamic element) => Text(element['name']),
          ),
        ),
      ),
    );

    expect(find.text('John'), findsOneWidget);
    expect(find.text('Danny'), findsOneWidget);
    expect(find.text('Team A'), findsOneWidget);
    expect(find.text('Team B'), findsOneWidget);
    expect(find.text('Team C'), findsWidgets);
  });

  testWidgets('empty list', (WidgetTester tester) async {
    await tester.pumpWidget(
      MaterialApp(
        home: Scaffold(
          body: StickyGroupedListView(
            groupBy: (dynamic element) => element['group'],
            elements: const [],
            groupSeparatorBuilder: buildGroupSeperator,
            itemBuilder: (context, dynamic element) => Text(element['name']),
          ),
        ),
      ),
    );
  });

  testWidgets('finds only one group separator per group',
      (WidgetTester tester) async {
    await tester.pumpWidget(
      MaterialApp(
        home: Scaffold(
          body: StickyGroupedListView(
            groupBy: (dynamic element) => element['group'],
            elements: _elements,
            groupSeparatorBuilder: buildGroupSeperator,
            itemBuilder: (context, dynamic element) => Text(element['name']),
          ),
        ),
      ),
    );
    expect(find.text("Team B"), findsOneWidget);
  });
}


example/lib/example.dart
import 'package:flutter/material.dart';
import 'package:sticky_grouped_list/sticky_grouped_list.dart';

/// This example shows the basic usage of the [StickyGroupedListView].
void main() => runApp(const MyApp());

List<Element> _elements = <Element>[
  Element(DateTime(2020, 6, 24, 18), 'Got to gym', Icons.fitness_center),
  Element(DateTime(2020, 6, 24, 9), 'Work', Icons.work),
  Element(DateTime(2020, 6, 25, 8), 'Buy groceries', Icons.shopping_basket),
  Element(DateTime(2020, 6, 25, 16), 'Cinema', Icons.movie),
  Element(DateTime(2020, 6, 25, 20), 'Eat', Icons.fastfood),
  Element(DateTime(2020, 6, 26, 12), 'Car wash', Icons.local_car_wash),
  Element(DateTime(2020, 6, 27, 12), 'Car wash', Icons.local_car_wash),
  Element(DateTime(2020, 6, 27, 13), 'Car wash', Icons.local_car_wash),
  Element(DateTime(2020, 6, 27, 14), 'Car wash', Icons.local_car_wash),
  Element(DateTime(2020, 6, 27, 15), 'Car wash', Icons.local_car_wash),
  Element(DateTime(2020, 6, 28, 12), 'Car wash', Icons.local_car_wash),
  Element(DateTime(2020, 6, 29, 12), 'Car wash', Icons.local_car_wash),
  Element(DateTime(2020, 6, 29, 12), 'Car wash', Icons.local_car_wash),
  Element(DateTime(2020, 6, 30, 12), 'Car wash', Icons.local_car_wash),
];

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Grouped List View Example'),
        ),
        body: StickyGroupedListView<Element, DateTime>(
          elements: _elements,
          order: StickyGroupedListOrder.ASC,
          groupBy: (Element element) => DateTime(
            element.date.year,
            element.date.month,
            element.date.day,
          ),
          groupComparator: (DateTime value1, DateTime value2) =>
              value2.compareTo(value1),
          itemComparator: (Element element1, Element element2) =>
              element1.date.compareTo(element2.date),
          floatingHeader: true,
          groupSeparatorBuilder: _getGroupSeparator,
          itemBuilder: _getItem,
        ),
      ),
    );
  }

  Widget _getGroupSeparator(Element element) {
    return SizedBox(
      height: 50,
      child: Align(
        alignment: Alignment.center,
        child: Container(
          width: 120,
          decoration: BoxDecoration(
            color: Colors.blue[300],
            border: Border.all(
              color: Colors.blue[300]!,
            ),
            borderRadius: const BorderRadius.all(Radius.circular(20.0)),
          ),
          child: Padding(
            padding: const EdgeInsets.all(8.0),
            child: Text(
              '${element.date.day}. ${element.date.month}, ${element.date.year}',
              textAlign: TextAlign.center,
            ),
          ),
        ),
      ),
    );
  }

  Widget _getItem(BuildContext ctx, Element element) {
    return Card(
      shape: RoundedRectangleBorder(
        borderRadius: BorderRadius.circular(6.0),
      ),
      elevation: 8.0,
      margin: const EdgeInsets.symmetric(horizontal: 10.0, vertical: 6.0),
      child: SizedBox(
        child: ListTile(
          contentPadding:
              const EdgeInsets.symmetric(horizontal: 20.0, vertical: 10.0),
          leading: Icon(element.icon),
          title: Text(element.name),
          trailing: Text('${element.date.hour}:00'),
        ),
      ),
    );
  }
}

class Element {
  DateTime date;
  String name;
  IconData icon;

  Element(this.date, this.name, this.icon);
}


example/lib/chat_example.dart
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';
import 'package:sticky_grouped_list/sticky_grouped_list.dart';

/// This example shows the usage of the [StickyGroupedListView] to create an
/// chat-like application where the elements are grouped by days and the
/// messages are shown on the left or right of the screen.
void main() => runApp(const MyApp());

List<Element> _elements = <Element>[
  Element(DateTime(2020, 6, 24, 9, 25), 'Hello how are you?'),
  Element(DateTime(2020, 6, 24, 9, 36), 'Fine and what about you?', true),
  Element(DateTime(2020, 6, 24, 9, 39), 'I am fine too'),
  Element(DateTime(2020, 6, 25, 14, 12),
      'Hey you do you wanna go to the cinema?', true),
  Element(
      DateTime(2020, 6, 25, 14, 19), 'Yes of course when do we want to meet'),
  Element(DateTime(2020, 6, 25, 14, 20), 'Lets meet at 8 o clock', true),
  Element(DateTime(2020, 6, 25, 14, 25), 'Okay see you then :)'),
  Element(DateTime(2020, 6, 27, 18, 41),
      'Hey whats up? Can you help me real quick?'),
  Element(DateTime(2020, 6, 27, 18, 45), 'Of course  what do you need?', true),
  Element(DateTime(2020, 6, 27, 18, 47),
      'Can you send me the homework for tomorrow please?'),
  Element(
    DateTime(2020, 6, 27, 18, 48),
    'I dont understand the math questions :(',
  ),
  Element(DateTime(2020, 6, 27, 18, 56), 'Yeah sure I have send them per mail',
      true),
];

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Chat with Peter'),
        ),
        body: Column(
          children: [
            SizedBox(
              height: 700,
              child: StickyGroupedListView<Element, DateTime>(
                elements: _elements,
                order: StickyGroupedListOrder.DESC,
                reverse: true,
                groupBy: (Element element) => DateTime(
                  element.date.year,
                  element.date.month,
                  element.date.day,
                ),
                floatingHeader: true,
                groupSeparatorBuilder: _getGroupSeparator,
                itemBuilder: _getItem,
              ),
            ),
            const TextField(
              decoration: InputDecoration(
                border: InputBorder.none,
                hintText: 'Enter a new message here',
              ),
            ),
          ],
        ),
      ),
    );
  }

  Widget _getGroupSeparator(Element element) {
    return SizedBox(
      height: 50,
      child: Align(
        alignment: Alignment.center,
        child: Container(
          width: 120,
          decoration: BoxDecoration(
            color: Colors.blue[300],
            border: Border.all(
              color: Colors.blue[300]!,
            ),
            borderRadius: const BorderRadius.all(Radius.circular(20.0)),
          ),
          child: Padding(
            padding: const EdgeInsets.all(8.0),
            child: Text(
              DateFormat.yMMMd().format(element.date),
              textAlign: TextAlign.center,
            ),
          ),
        ),
      ),
    );
  }

  Widget _getItem(BuildContext ctx, Element element) {
    return Align(
      alignment: element.swapped ? Alignment.centerRight : Alignment.centerLeft,
      child: SizedBox(
        width: 370,
        child: Card(
          shape: RoundedRectangleBorder(
            borderRadius: BorderRadius.circular(6.0),
          ),
          elevation: 8.0,
          margin: const EdgeInsets.symmetric(horizontal: 10.0, vertical: 6.0),
          child: SizedBox(
            child: ListTile(
              contentPadding:
                  const EdgeInsets.symmetric(horizontal: 20.0, vertical: 10.0),
              leading: element.swapped
                  ? Text(DateFormat.Hm().format(element.date))
                  : const Icon(Icons.person),
              title: Text(element.name),
              trailing: element.swapped
                  ? const Icon(Icons.person_outline)
                  : Text(DateFormat.Hm().format(element.date)),
            ),
          ),
        ),
      ),
    );
  }
}

class Element implements Comparable {
  DateTime date;
  String name;
  bool swapped = false;

  Element(this.date, this.name, [this.swapped = false]);

  @override
  int compareTo(other) {
    return date.compareTo(other.date);
  }
}


example/lib/scroll_example.dart
import 'package:flutter/material.dart';
import 'package:sticky_grouped_list/sticky_grouped_list.dart';

/// This example shows the usage of the [StickyGroupedListView] with a
/// [GroupedItemScrollController]. The list shows 100 elements divided in ten
/// groups.
/// When clicking the [FloatingActionButton] the scroll controller jumps to the
/// index 50.
void main() => runApp(MyApp());

List<Element> _elements =
    List.generate(100, (index) => Element(index, 'Item $index', index ~/ 10));

class MyApp extends StatelessWidget {
  MyApp({Key? key}) : super(key: key);

  final itemScrollController = GroupedItemScrollController();
  final itemPositionsListener = ItemPositionsListener.create();

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Grouped List View Example'),
        ),
        body: StickyGroupedListView<Element, int>(
          elements: _elements,
          groupBy: (Element element) => element.group,
          elementIdentifier: (Element element) => element.id,
          floatingHeader: true,
          groupSeparatorBuilder: _getGroupSeparator,
          itemBuilder: _getItem,
          itemPositionsListener: itemPositionsListener,
          initialScrollIndex: 10,
          itemScrollController: itemScrollController,
        ),
        floatingActionButton: FloatingActionButton(
          child: const Icon(Icons.arrow_forward),
          onPressed: () {
            itemScrollController.jumpToElement(
              identifier: 50,
            );
          },
        ),
      ),
    );
  }

  Widget _getGroupSeparator(Element element) {
    return SizedBox(
      height: 50,
      child: Align(
        alignment: Alignment.center,
        child: Container(
          width: 120,
          decoration: BoxDecoration(
            color: Colors.blue[300],
            border: Border.all(
              color: Colors.blue[300]!,
            ),
            borderRadius: const BorderRadius.all(Radius.circular(20.0)),
          ),
          child: Padding(
            padding: const EdgeInsets.all(8.0),
            child: Text('${element.group}', textAlign: TextAlign.center),
          ),
        ),
      ),
    );
  }

  Widget _getItem(BuildContext ctx, Element element) {
    return Card(
      shape: RoundedRectangleBorder(
        borderRadius: BorderRadius.circular(6.0),
      ),
      elevation: 8.0,
      margin: const EdgeInsets.symmetric(horizontal: 10.0, vertical: 6.0),
      child: SizedBox(
        child: ListTile(
          contentPadding:
              const EdgeInsets.symmetric(horizontal: 20.0, vertical: 10.0),
          title: Text(element.name),
        ),
      ),
    );
  }
}

class Element {
  int id;
  String name;
  int group;

  Element(this.id, this.name, this.group);
}


example/lib/chatExample.dart
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';
import 'package:sticky_grouped_list/sticky_grouped_list.dart';

void main() => runApp(MyApp());

List<Element> _elements = <Element>[
  Element(DateTime(2020, 6, 24, 9, 25), 'Hello how are you?'),
  Element(DateTime(2020, 6, 24, 9, 36), 'Fine and what about you?', true),
  Element(DateTime(2020, 6, 24, 9, 39), 'I am fine too'),
  Element(DateTime(2020, 6, 25, 14, 12),
      'Hey you do you wanna go to the cinema?', true),
  Element(
      DateTime(2020, 6, 25, 14, 19), 'Yes of course when do we want to meet'),
  Element(DateTime(2020, 6, 25, 14, 20), 'Lets meet at 8 o clock', true),
  Element(DateTime(2020, 6, 25, 14, 25), 'Okay see you then :)'),
  Element(DateTime(2020, 6, 27, 18, 41),
      'Hey whats up? Can you help me real quick?'),
  Element(DateTime(2020, 6, 27, 18, 45), 'Of course  what do you need?', true),
  Element(DateTime(2020, 6, 27, 18, 47),
      'Can you send me the homework for tomorrow please?'),
  Element(
    DateTime(2020, 6, 27, 18, 48),
    'I dont understand the math questions :(',
  ),
  Element(DateTime(2020, 6, 27, 18, 56), 'Yeah sure I have send them per mail',
      true),
];

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        appBar: AppBar(
          title: Text('Chat with Peter'),
        ),
        body: Column(
          children: [
            Container(
              height: 700,
              child: StickyGroupedListView<Element, DateTime>(
                elements: _elements,
                order: StickyGroupedListOrder.DESC,
                reverse: true,
                groupBy: (Element element) => DateTime(
                    element.date.year, element.date.month, element.date.day),
                floatingHeader: true,
                groupSeparatorBuilder: (Element element) => Container(
                  height: 50,
                  child: Align(
                    alignment: Alignment.center,
                    child: Container(
                      width: 120,
                      decoration: BoxDecoration(
                        color: Colors.blue[300],
                        border: Border.all(
                          color: Colors.blue[300]!,
                        ),
                        borderRadius: BorderRadius.all(Radius.circular(20.0)),
                      ),
                      child: Padding(
                        padding: const EdgeInsets.all(8.0),
                        child: Text(
                          '${DateFormat.yMMMd().format(element.date)}',
                          textAlign: TextAlign.center,
                        ),
                      ),
                    ),
                  ),
                ),
                itemBuilder: (_, Element element) {
                  return Align(
                    alignment: element.swapped
                        ? Alignment.centerRight
                        : Alignment.centerLeft,
                    child: Container(
                      width: 370,
                      child: Card(
                        shape: RoundedRectangleBorder(
                          borderRadius: BorderRadius.circular(6.0),
                        ),
                        elevation: 8.0,
                        margin: new EdgeInsets.symmetric(
                            horizontal: 10.0, vertical: 6.0),
                        child: Container(
                          child: ListTile(
                            contentPadding: EdgeInsets.symmetric(
                                horizontal: 20.0, vertical: 10.0),
                            leading: element.swapped
                                ? Text(DateFormat.Hm().format(element.date))
                                : Icon(Icons.person),
                            title: Text(element.name),
                            trailing: element.swapped
                                ? Icon(Icons.person_outline)
                                : Text(DateFormat.Hm().format(element.date)),
                          ),
                        ),
                      ),
                    ),
                  );
                },
              ),
            ),
            TextField(
              decoration: InputDecoration(
                  border: InputBorder.none,
                  hintText: 'Enter a new message here'),
            ),
          ],
        ),
      ),
    );
  }
}

class Element implements Comparable {
  DateTime date;
  String name;
  bool swapped = false;

  Element(this.date, this.name, [this.swapped = false]);

  @override
  int compareTo(other) {
    return date.compareTo(other.date);
  }
}
